Source/DafnyCore/Resolver/PreType/PreTypeResolve.State.cs

```diff
@@ -0,0 +1,589 @@
+      if (!options.Get(CommonOptionBag.NewTypeInferenceDebug)) {
+        return;
+      }
+      options.OutputWriter.WriteLine("*** Type inference state ***{0}", header == null ? "" : $" {header} ");
+      PrintList("Subtype constraints", unnormalizedSubtypeConstraints, stc => {
+        return $"{stc.Super} :> {stc.Sub}";
+      });
+      PrintList("Equality constraints", equalityConstraints, eqc => {
+        return $"{eqc.A} == {eqc.B}";
+      });
+#if IS_THERE_A_GOOD_WAY_TO_PRINT_GUARDED_CONSTRAINTS
+      PrintList("Guarded constraints", guardedConstraints, gc => {
+        return gc.Kind + Util.Comma("", gc.Arguments, arg => $" {arg}");
+      });
+#endif
+      PrintList("Default-type advice", defaultAdvice, advice => {
+        return $"{advice.PreType} ~-~-> {advice.WhatString}";
+      });
+#if IS_THERE_A_GOOD_WAY_TO_PRINT_CONFIRMATIONS
+      PrintList("Post-inference confirmations", confirmations, c => {
+        return $"{TokToShortLocation(c.tok)}: {c.Check} {c.PreType}: {c.ErrorMessage()}";
+      });
+#endif
+    }
+
+    void PrintLegend() {
+      PrintList("Legend", PreTypeResolver.allPreTypeProxies, pair => {
+        var s = Pad($"?{pair.Item1.UniqueId}", 4) + pair.Item1;
+        return pair.Item2 == null ? s : $"{Pad(s, 20)}  {pair.Item2}";
+      });
+    }
+
+    void PrintList<T>(string rubric, List<T> list, Func<T, string> formatter) {
+      if (!options.Get(CommonOptionBag.NewTypeInferenceDebug)) {
+        return;
+      }
+      options.OutputWriter.WriteLine($"    {rubric}:");
+      foreach (var t in list) {
+        var info = $"        {formatter(t)}";
+        if (t is PreTypeStateWithErrorMessage preTypeStateWithErrorMessage) {
+          info = $"{Pad(info, 30)}  {TokToShortLocation(preTypeStateWithErrorMessage.tok)}: {preTypeStateWithErrorMessage.ErrorMessage()}";
+        }
+        options.OutputWriter.WriteLine(info);
+      }
+    }
+
+    public void DebugPrint(string format, params object[] args) {
+      if (options.Get(CommonOptionBag.NewTypeInferenceDebug)) {
+        Console.WriteLine(format, args);
+      }
+    }
+
+    // ---------------------------------------- Equality constraints ----------------------------------------
+
+    public void AddEqualityConstraint(PreType a, PreType b, IToken tok, string msgFormat) {
+      equalityConstraints.Add(new EqualityConstraint(a, b, tok, msgFormat));
+    }
+
+    private bool ApplyEqualityConstraints() {
+      if (equalityConstraints.Count == 0) {
+        return false;
+      }
+      var constraints = equalityConstraints;
+      equalityConstraints = new();
+      foreach (var constraint in constraints) {
+        equalityConstraints.AddRange(constraint.Apply(this));
+      }
+      return true;
+    }
+
+    // ---------------------------------------- Comparable constraints ----------------------------------------
+
+    public void AddComparableConstraint(PreType a, PreType b, IToken tok, string errorFormatString) {
```

Comment:
I didn't understand your reply to my previous comment, but I don't think this code belongs in this file.

Reply:
I removed the code.